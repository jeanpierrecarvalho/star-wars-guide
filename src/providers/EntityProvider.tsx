import React, { useState, useCallback, useContext } from 'react'
import { EntityContext } from '@/contexts/EntityContext'
import { fetchEntity } from '@/services/star-wars-api'
import { ENTITY_TYPE } from '@/constants/entities'

interface IEntity {
  [key: string]: any
}

export const useEntityContext = () => {
  const context = useContext(EntityContext)
  if (!context) {
    throw new Error('useEntityContext must be used within an EntityProvider')
  }
  return context
}

export const EntityProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [data, setData] = useState<IEntity[]>([])
  const [page, setPage] = useState<number>(1)
  const [totalPages, setTotalPages] = useState<number>(1)
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [loading, setLoading] = useState<boolean>(false)
  const [error, setError] = useState<string | null>(null)

  const loadData = async (type: string) => {
    setLoading(true)
    try {
      const response = await fetchEntity(ENTITY_TYPE[type].toFetch, page, searchTerm)
      setData(response.results)
      setTotalPages(Math.ceil(response.count / 10))
    } catch (err) {
      setError('Failed to load data. Please try again later.')
    } finally {
      setLoading(false)
    }
  }

  const handlePageChange = useCallback(
    (newPage: number) => {
      if (newPage > 0 && newPage <= totalPages) {
        setPage(newPage)
      }
    },
    [totalPages],
  )

  const handleSearchChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(e.target.value)
    setPage(1)
  }, [])

  return (
    <EntityContext.Provider
      value={{
        data,
        page,
        totalPages,
        searchTerm,
        loading,
        error,
        loadData,
        handlePageChange,
        handleSearchChange,
      }}
    >
      {children}
    </EntityContext.Provider>
  )
}
