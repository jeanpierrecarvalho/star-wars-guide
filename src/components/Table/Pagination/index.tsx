import { useEntityContext } from "@/providers/EntityProvider";

const Pagination: React.FC = () => {
  const { page, totalPages, handlePageChange } = useEntityContext();

  return (
    <div className="flex flex-row justify-between items-center">
      <button
        className={`${
          page === 1
            ? "bg-yellow-500 opacity-40 cursor-not-allowed"
            : "bg-yellow-500 hover:bg-yellow-300"
        } px-2 py-1 rounded-lg w-32 font-medium text-gray-900`}
        onClick={() => handlePageChange(page - 1)}
        disabled={page === 1}
        aria-label="Previous Page"
        aria-disabled={page === 1}
      >
        Previous
      </button>
      <span className="flex flex-row items-center text-center text-lg">
        Page {page} of {totalPages}
      </span>
      <button
        className={`${
          page === totalPages
            ? "bg-yellow-500 opacity-40 cursor-not-allowed"
            : "bg-yellow-500 hover:bg-yellow-300"
        } px-2 py-1 rounded-lg w-32 font-medium text-gray-900`}
        onClick={() => handlePageChange(page + 1)}
        disabled={page === totalPages}
        aria-label="Next Page"
        aria-disabled={page === totalPages}
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
