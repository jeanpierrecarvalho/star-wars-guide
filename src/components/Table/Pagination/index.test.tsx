import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import Pagination from "@/components/Table/Pagination";

import { EntityProvider } from "@/providers/EntityProvider";

describe("Pagination Component", () => {
  it("matches the snapshot", () => {
    const { container } = render(
      <EntityProvider>
        <Pagination />
      </EntityProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
