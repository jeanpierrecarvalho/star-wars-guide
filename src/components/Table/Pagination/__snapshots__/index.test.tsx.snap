// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination Component > matches the snapshot 1`] = `
<div>
  <div
    class="flex flex-row justify-between items-center"
  >
    <button
      aria-disabled="true"
      aria-label="Previous Page"
      class="bg-yellow-500 opacity-40 cursor-not-allowed px-2 py-1 rounded-lg w-32 font-medium text-gray-900"
      disabled=""
    >
      Previous
    </button>
    <span
      class="flex flex-row items-center text-center text-lg"
    >
      Page 
      1
       of 
      1
    </span>
    <button
      aria-disabled="true"
      aria-label="Next Page"
      class="bg-yellow-500 opacity-40 cursor-not-allowed px-2 py-1 rounded-lg w-32 font-medium text-gray-900"
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
